lib/Heartbeat.js

```diff
@@ -0,0 +1,248 @@
+const {Cu} = require('chrome');
+
+Cu.import('resource://gre/modules/Services.jsm'); /* globals Services */
+Cu.import('resource://gre/modules/TelemetryController.jsm'); /* globals TelemetryController */
+Cu.import('resource://gre/modules/Timer.jsm'); /* globals setTimeout, clearTimeout */
+
+Cu.importGlobalProperties(['URL']); /* globals URL */
+
+const {Log} = require('./Log.js');
+const {EventEmitter} = require('./EventEmitter.js');
+
+const PREF_SURVEY_DURATION = 'browser.uitour.surveyDuration';
+
+/**
+ * Show the Heartbeat UI to request user feedback. This function reports back to the
+ * caller using |notify|. The notification event name reflects the current status the UI
+ * is in (either "Heartbeat:NotificationOffered", "Heartbeat:NotificationClosed",
+ * "Heartbeat:LearnMore", "Heartbeat:Engaged", "Heartbeat:Voted",
+ * "Heartbeat:SurveyExpired" or "Heartbeat:WindowClosed").
+ * When a "Heartbeat:Voted" event is notified
+ * the data payload contains a |score| field which holds the rating picked by the user.
+ * Please note that input parameters are already validated by the caller.
+ *
+ * @param chromeWindow
+ *        The chrome window that the heartbeat notification is displayed in.
+ * @param {Object} options Options object.
+ * @param {String} options.message
+ *        The message, or question, to display on the notification.
+ * @param {String} options.thankyouMessage
+ *        The thank you message to display after user votes.
+ * @param {String} options.flowId
+ *        An identifier for this rating flow. Please note that this is only used to
+ *        identify the notification box.
+ * @param {String} [options.engagementButtonLabel=null]
+ *        The text of the engagement button to use instad of stars. If this is null
+ *        or invalid, rating stars are used.
+ * @param {String} [options.engagementURL=null]
+ *        The engagement URL to open in a new tab once user has engaged. If this is null
+ *        or invalid, no new tab is opened.
+ * @param {String} [options.learnMoreLabel=null]
+ *        The label of the learn more link. No link will be shown if this is null.
+ * @param {String} [options.learnMoreURL=null]
+ *        The learn more URL to open when clicking on the learn more link. No learn more
+ *        will be shown if this is an invalid URL.
+ * @param {boolean} [options.privateWindowsOnly=false]
+ *        Whether the heartbeat UI should only be targeted at a private window (if one exists).
+ *        No notifications should be fired when this is true.
+ * @param {String} [options.surveyId]
+ *        An ID for the survey, reflected in the Telemetry ping.
+ * @param {Number} [options.surveyVersion]
+ *        Survey's version number, reflected in the Telemetry ping.
+ * @param {boolean} [options.testing]
+ *        Whether this is a test survey, reflected in the Telemetry ping.
+ */
+exports.Heartbeat = class {
```

Comment:
Do we need to handle this `privateWindowsOnly` option? https://dxr.mozilla.org/mozilla-central/rev/f97a056ae6235de7855fd8aaa04fb1c8d183bd06/browser/components/uitour/UITour.jsm#1108-1111